### ISSUE TITLE
ttnn.subtract and ttnn.remainder do not support implicit broadcast on both operands

### ISSUE THREAD

**Issue by jserbedzijaTT** [tenstorrent/tt-metal#24635]
### Describe the bug

ttnn.subtract and ttnn.remainder operations do not support broadcast on both operands

### Steps to reproduce the issue

The following test captures the failing scenarios:

```
import pytest

import torch

import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

@pytest.mark.parametrize(
    "shapes",
    [
        [[1, 1, 1], [8, 16, 32]],
        [[1, 1, 32], [8, 16, 32]],
    ],
)
def test_sub_implicit_broadcast(device, shapes):
    torch.manual_seed(0)

    torch_input_tensor_a = torch.rand(shapes[0], dtype=torch.float32)
    torch_input_tensor_b = torch.rand(shapes[1], dtype=torch.float32)
    torch_output_tensor = torch.sub(torch_input_tensor_a, torch_input_tensor_b)

    input_tensor_a = ttnn.from_torch(
        torch_input_tensor_a, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    input_tensor_b = ttnn.from_torch(
        torch_input_tensor_b, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )

    output_tensor = ttnn.sub(input_tensor_a, input_tensor_b, dtype=ttnn.float32, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor, 0.9999)

@pytest.mark.parametrize(
    "shapes",
    [
        [[1, 1, 1], [8, 16, 32]],
        [[1, 1, 32], [8, 16, 32]],
        [[1, 16, 32], [8, 16, 32]],
        [[8, 16, 1], [1, 1, 32]],
        [[1, 1, 32], [8, 16, 1]],
        [[8, 1, 32], [8, 16, 1]],
        [[8, 16, 1], [8, 1, 32]],
    ],
)
def test_remainder_implicit_broadcast(device, shapes):
    torch.manual_seed(0)

    torch_input_tensor_a = torch.rand(shapes[0], dtype=torch.float32)
    torch_input_tensor_b = torch.rand(shapes[1], dtype=torch.float32)
    torch_output_tensor = torch.remainder(torch_input_tensor_a, torch_input_tensor_b)

    input_tensor_a = ttnn.from_torch(
        torch_input_tensor_a, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    input_tensor_b = ttnn.from_torch(
        torch_input_tensor_b, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )

    output_tensor = ttnn.remainder(input_tensor_a, input_tensor_b, dtype=ttnn.float32, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor, 0.9999)
```

**Comment by cmaryanTT** [comment]
All of these cases should be supported, but may be affected by the recent revert. @dchenTT - can you please check.

**Comment by dchenTT** [comment]
For first test ttnn.sub, both shapes are directed to legacy implementation. The first shape is tensor-scalar, tensor-scalar is currently directed to legacy because the revert. The 2nd shape is subtile row broadcast (on h dimension) that is also directed to legacy due to performance issue. Therefore, for both shapes, once they are handled by legacy, the broadcasted operand have to be the 2nd operand. So in order to pass the test, either swap the first and second operand, or add flag use_legacy=False to explicitly directed to binary_ng implementation. In the future, the problem will be fixed without user intervention.

For 2nd test ttnn.remainder, reason is different. Still investigating.

**Comment by dchenTT** [comment]
Some findings for reminder op:
1. The op does not support output parameter such as dtype=
2. It does not support broadcasting well.  If tensors a (the first operand) has the final broadcasted shape, it is OK.

**Comment by dchenTT** [comment]
PR opened for ttnn.remainder https://github.com/tenstorrent/tt-metal/pull/25224

**Comment by dchenTT** [comment]
@jserbedzijaTT it has been merged a few days ago. Could you verify and close?

**Comment by jserbedzijaTT** [comment]
There are still some tests that fail with the Int32 dtype, as shown in the following test:
```
@pytest.mark.parametrize(
     "shapes",
     [
         # no subtile bcast
         [[1, 16, 32], [8, 16, 32]],
         # scalar bcast
         [[8, 16, 32], [1, 1, 1]],
         [[1, 1, 1], [8, 16, 32]],
         # col bcast
         [[1, 16, 1], [8, 16, 32]],
         [[8, 16, 32], [8, 16, 1]],
         # row bcast
         [[8, 16, 32], [8, 1, 32]],
         [[8, 1, 32], [8, 16, 32]],
         # row col mixed bcast
         [[1, 1, 32], [8, 16, 1]],
         [[8, 16, 1], [1, 1, 32]],
     ],
 )
def test_remainder_implicit_broadcast(device, shapes):
    torch.manual_seed(0)

    torch_input_tensor_a = torch.randint(
        torch.iinfo(torch.int32).min,
        torch.iinfo(torch.int32).max,
        shapes[0],
        dtype=torch.int32
    )

    torch_input_tensor_b = torch.randint(
        torch.iinfo(torch.int32).min,
        torch.iinfo(torch.int32).max,
        shapes[1],
        dtype=torch.int32
    )

    torch_output_tensor = torch.remainder(torch_input_tensor_a, torch_input_tensor_b)

    input_tensor_a = ttnn.from_torch(
        torch_input_tensor_a,
        dtype=ttnn.float32,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )
    input_tensor_b = ttnn.from_torch(
        torch_input_tensor_b,
        dtype=ttnn.float32,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )

    output_tensor = ttnn.remainder(input_tensor_a, input_tensor_b, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    output_tensor = ttnn.to_torch(output_tensor)
    assert_with_pcc(torch_output_tensor, output_tensor, 0.9999)
```

**Comment by fahmedchanTT** [comment]
@dchenTT What are the next steps here ?

**Comment by dchenTT** [comment]
I can reproduce the failed cases. What it is puzzling to me is that for int32, the op will typcast input to float32 anyway then downcast on output. So in theory, treating float32 and int32 as input should have the same workflow, however, the debug log shows differently. Since remainder is a composite op with many individual ops plus typecasting, the scope is not small. I am in progress to debug it.

**Comment by dchenTT** [comment]
This issue is first related to typcast handling as I mentioned above. I did add a workaround to force to use binary_ng for ttnn::multiply, and it worked better, as I observed that for some reason (may be because typecasting), the failed case choose legacy and passed case choose binary_ng . The second issue is about the accuracy of remainder for int32 when casting to float32. When int32 value is small like [-1000, 1000], the test can pass. When int32 value becomes bigger, the accuracy will lose more to fail, not small margin. For both issues, I think it is better for @KalaivaniMCW to take a look.

**Comment by mcw-anasuya** [comment]
Failed ttnn.remainder cases is a broadcast issue. Row bcast and col-row mixed bcast fail when the second tensor is being broadcasted for large int32 and fp32 inputs. Sub-ops multiply, add and subtract were calling broadcast height program factory instead of binary_ng. Setting use_legacy = false fixed the issue. 
Fix PR: #26338

**Comment by KalaivaniMCW** [comment]
This is the op flow for remainder and for ADD/SUB/MUL with row-bcast we fallback to legacy bcast kernels
Looks like there are some unidentified issues with bcast kernels. Will do a layer-by-layer debug 
But using the binary_ng kernels work. 

<img width="928" height="652" alt="Image" src="https://github.com/user-attachments/assets/c4e7271f-e507-4cce-a757-dd58ed8a3296" />

**Comment by dchenTT** [comment]
I thought I fixed the float32 version, only int32 had problem. But it turned out float32 can also fail. The reason why it did not show up is by default torch.rand() generates float value [0.0, 1.0]. The range is small so it hides problem. By using a bigger range, it will fail too. So the problem is not about float32 vs int32, and not about typecasting. It is about math/bcast problem in some legacy op. Since by direct everything to binary_ng fixed the problem. @KalaivaniMCW I recommend we do not spend time debugging legacy code.

**Comment by KalaivaniMCW** [comment]
Noted @dchenTT

**Comment by mcw-anasuya** [comment]
@jserbedzijaTT Fix for ttnn.remainder has been merged to main. Please verify.

**Comment by KalaivaniMCW** [comment]
Closing this as fix has been merged. Pls reopen if you face any problems related to this.
